### ISSUE TITLE
DNS provider bug. DNS should not fail when record does not exist.

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-dns#2]
_This issue was originally opened by @vmindru as hashicorp/terraform#15088. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

terraform should not fail when the record does not exist, it should create the record instead. 

### Terraform Version

Terraform v0.9.6

### Affected Resource(s)

dns_a_record_set

### Terraform Configuration Files

```
[vmindru@vmutil test]$ cat dns.tf 
provider "dns" {
  update {
    server        = "${ var.dnsmaster }"
    key_name      = "${ var.dnskey }."
    key_algorithm = "hmac-md5"
    key_secret    = "${ var.dnskeysecret }"
  }
}
[vmindru@vmutil test]$ cat dns-zone.tf 
resource "dns_a_record_set" "www" {
  zone = "zone.com."
  name = "test2"
  addresses = [ "127.0.0.1" ]
  ttl = 30
}
[vmindru@vmutil test]$ 
```

### Debug Output

trucnated 

```

dns_a_record_set.www - *terraform.NodeRefreshableManagedResourceInstance
2017/06/05 13:04:43 [DEBUG] dag/walk: added new vertex: "dns_a_record_set.www"
2017/06/05 13:04:43 [DEBUG] dag/walk: walking "dns_a_record_set.www"
2017/06/05 13:04:43 [DEBUG] vertex 'root.dns_a_record_set.www': walking
2017/06/05 13:04:43 [DEBUG] vertex 'root.dns_a_record_set.www': evaluating
2017/06/05 13:04:43 [TRACE] [walkRefresh] Entering eval tree: dns_a_record_set.www
2017/06/05 13:04:43 [DEBUG] root: eval: *terraform.EvalSequence
2017/06/05 13:04:43 [DEBUG] root: eval: *terraform.EvalGetProvider
2017/06/05 13:04:43 [DEBUG] root: eval: *terraform.EvalReadState
2017/06/05 13:04:43 [DEBUG] root: eval: *terraform.EvalRefresh
2017/06/05 13:04:44 [ERROR] root: eval: *terraform.EvalRefresh, err: dns_a_record_set.www: Error querying DNS record: 3
2017/06/05 13:04:44 [ERROR] root: eval: *terraform.EvalSequence, err: dns_a_record_set.www: Error querying DNS record: 3
Error refreshing state: 1 error(s) occurred:
2017/06/05 13:04:44 [DEBUG] plugin: waiting for all plugin processes to complete...
2017/06/05 13:04:44 [DEBUG] plugin: terraform: dns-provider (internal) 2017/06/05 13:04:44 [DEBUG] plugin: waiting for all plugin processes to complete...

* dns_a_record_set.www: 1 error(s) occurred:

* dns_a_record_set.www: dns_a_record_set.www: Error querying DNS record: 3

```

### Expected Behavior

if DNS record does not exist the zone should be created hence it does not exist. 

### Actual Behavior

terraform end's with an error 

### Steps to Reproduce
 `terraform apply`

### Important Factoids

With my limited go understanding, I think the hole problem is that terraform fail's based on the logic: "
fail on any RCode Value except 0" 
This is behaviour is not correct, RCode 3 is a normal error that means that the queried record does not exist. 

http://www.tcpipguide.com/free/t_DNSMessageHeaderandQuestionSectionFormat.htm

**Comment by apparentlymart** [comment]
Sorry for the delayed reply here @vmindru.

Agreed that Terraform should ideally handle an `NXDOMAIN` (error code 3) during refresh by removing the resource from the state, just as we do when other resources show as deleted. Generally speaking it's not recommended to make changes outside of Terraform to resources Terraform is managing, but Terraform _should_ detect such changes and plan to fix them, rather than producing an error as in this case.

**Comment by vmindru** [comment]
yep i agree the resource should be removed from the state file and then created.

**Comment by bpoland** [comment]
Is anyone working on a fix for this?

**Comment by bpoland** [comment]
@apparentlymart do you know if there are any plans to fix this issue?  Thanks.

**Comment by vmindru** [comment]
@bpoland  i lost hope :-D ike 3 months ago :-D

**Comment by apparentlymart** [comment]
The Terraform team at HashiCorp won't be able to work on this in the near future due to our focus being elsewhere, but we'd be happy to review a pull request if you or someone else has the time and motivation to implement it.

**Comment by bpoland** [comment]
Okay, thank you for letting us know.  Unfortunately I have no experience with go myself -- I was hoping it would be a relatively easy fix for someone that's already familiar with the provider.

@bodgit I notice you have been contributing to this provider recently -- thank you for that!  Is there any chance you would have some time to take a look at this issue?  Thanks!

**Comment by bodgit** [comment]
I can take a look, if I understand the problem correctly.

**Comment by bpoland** [comment]
That would be awesome.  Please let us know if anything is unclear.  I think @apparentlymart summarized it best above:

> Terraform should ideally handle an NXDOMAIN (error code 3) during refresh by removing the resource from the state, just as we do when other resources show as deleted. Generally speaking it's not recommended to make changes outside of Terraform to resources Terraform is managing, but Terraform should detect such changes and plan to fix them, rather than producing an error as in this case.

**Comment by bodgit** [comment]
I've created #33 which I think fixes this.

**Comment by bpoland** [comment]
Great, thanks so much @bodgit!  Really appreciate your time and effort :)

**Comment by katbyte** [comment]
Hi @vmindru,

Just wanted to let you know that we have released [v2.0.0](https://github.com/terraform-providers/terraform-provider-dns/blob/master/CHANGELOG.md#200-may-25-2018) of the dns provider that should no longer fail when the DNS record doesn't exist 🙂
